feat: add fixed position coordinate picker

dzienisz opened the pull request:
## Description

<img width="1208" height="547" alt="Screenshot 2025-11-14 at 14 40 01" src="https://github.com/user-attachments/assets/a5eaa9f3-769a-415f-83e7-bff50e62a99f" />
<img width="1240" height="559" alt="Screenshot 2025-11-14 at 14 40 05" src="https://github.com/user-attachments/assets/3f0d45d1-2f39-4526-8ea0-3de3e152acb5" />

# Integrate Fixed Position Admin Messages into Global Save

## Summary
Refactored the fixed position saving mechanism to integrate admin message queuing into the global save workflow. Previously, the [setFixedPosition](cci:1://file:///Users/home/Developer/web/packages/core/src/meshDevice.ts:385:2-410:3) admin message was sent immediately when the form changed, causing a race condition where the admin message could arrive before or after the `fixedPosition: true` config was committed. Now, admin messages are queued alongside config changes and sent in the correct order during the global save operation.

## Related Issues
Fixes: Fixed position coordinates were not reliably applied due to race conditions between immediate admin message sending and deferred config commits via the global save button.

## Changes Made

### New Infrastructure
- **[changeRegistry.ts](cci:7://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:0:0-0:0)**
  - Extended [ConfigChangeKey](cci:2://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:31:0-36:73) type to include `adminMessage` variant with [ValidAdminMessageType](cci:2://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:28:0-28:65) and unique ID.
  - Updated [serializeKey](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:54:0-70:1) and [deserializeKey](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:72:0-103:1) to handle admin message keys.
  - Added [getAllAdminMessages()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:234:0-245:1) and [getAdminMessageChangeCount()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:247:0-259:1) helper functions.

- **[deviceStore/index.ts](cci:7://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:0:0-0:0)**
  - Added [queueAdminMessage()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:948:4-977:5), [getAllQueuedAdminMessages()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:979:4-989:5), and [getAdminMessageChangeCount()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:247:0-259:1) methods to the [Device](cci:2://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:62:0-153:1) interface.
  - Implemented admin message queuing in [deviceFactory](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:191:0-1000:1) using the unified change registry.

### Updated Files
- **[Position.tsx](cci:7://file:///Users/home/Developer/web/packages/web/src/components/PageComponents/Settings/Position.tsx:0:0-0:0)**
  - Replaced [sendAdminMessage](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:690:4-701:5) with [queueAdminMessage](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:948:4-977:5) for [setFixedPosition](cci:1://file:///Users/home/Developer/web/packages/core/src/meshDevice.ts:385:2-410:3) admin messages.
  - Admin messages are now queued during form submission instead of sent immediately.
  - Removed immediate position update request (no longer needed).

- **[Settings/index.tsx](cci:7://file:///Users/home/Developer/web/packages/web/src/pages/Settings/index.tsx:0:0-0:0)**
  - Added `toBinary` and `Protobuf` imports.
  - Integrated [getAllQueuedAdminMessages()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:979:4-989:5) and [getAdminMessageChangeCount()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/changeRegistry.ts:247:0-259:1) into the component.
  - Updated `handleSave` to process queued admin messages **after** configs are committed.
  - Admin message changes now count toward pending changes indicator.
  - Added null check for `ActiveComponent` to prevent TypeScript errors.

## Key Points
- **Correct Sequencing**: Admin messages are sent after [commitEditSettings()](cci:1://file:///Users/home/Developer/web/packages/core/src/meshDevice.ts:594:2-609:3), ensuring the device has the correct configuration before receiving position coordinates.
- **Unified Change Tracking**: Admin messages use the same change registry system as configs, modules, and channels.
- **No Race Conditions**: The `fixedPosition: true` config is guaranteed to be committed before the [setFixedPosition](cci:1://file:///Users/home/Developer/web/packages/core/src/meshDevice.ts:385:2-410:3) admin message is sent.
- **Clean Integration**: Admin messages are treated as first-class changes in the global save workflow.
- **Type Safety**: Full TypeScript support with proper schemas and validation.

## Technical Details

### Admin Message Flow
1. User edits position coordinates in the form
2. [onSubmit](cci:1://file:///Users/home/Developer/web/packages/web/src/components/PageComponents/Settings/Position.tsx:60:2-134:4) in Position.tsx is triggered (via `onChange`)
3. Config changes are queued via [setChange()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:759:4-781:5)
4. Admin message is queued via [queueAdminMessage()](cci:1://file:///Users/home/Developer/web/packages/web/src/core/stores/deviceStore/index.ts:948:4-977:5)
5. User clicks global Save button
6. `handleSave` processes changes in order:
   - Sends channel changes
   - Sends config changes (including `fixedPosition: true`)
   - Sends module config changes
   - **Commits settings** ([commitEditSettings()](cci:1://file:///Users/home/Developer/web/packages/core/src/meshDevice.ts:594:2-609:3))
   - **Sends queued admin messages** ([setFixedPosition](cci:1://file:///Users/home/Developer/web/packages/core/src/meshDevice.ts:385:2-410:3))
   - Updates local state
   - Clears all changes

### Change Registry Schema
```typescript
type ConfigChangeKey =
  | { type: "config"; variant: ValidConfigType }
  | { type: "moduleConfig"; variant: ValidModuleConfigType }
  | { type: "channel"; index: Types.ChannelNumber }
  | { type: "user" }
  | { type: "adminMessage"; variant: ValidAdminMessageType; id: string };
```

## Testing
- ✅ Lint: `pnpm run lint` (root) — passes.
- ✅ Type checking: No TypeScript errors.
- ✅ Manual verification:
  - Fixed position coordinates are correctly queued when form changes.
  - Global save button appears when admin messages are queued.
  - Admin messages are sent after config commit.
  - No race conditions observed.
  - Change count includes admin messages.

## Checklist
- [x] Code follows project style/patterns.
- [x] Type safety maintained (proper schemas and interfaces).
- [x] Unified change tracking system extended.
- [x] Linting passes.
- [x] No console.log statements in production code.
- [x] Null safety checks added where needed.

danditomaso commented on packages/core/src/utils/transform/decodePacket.ts:
```diff
@@ -194,7 +194,7 @@ export const decodePacket = (device: MeshDevice) =>
               device.log.trace(
                 Types.Emitter[Types.Emitter.HandleFromRadio],
-                `🚧 Received Queue Status: ${decodedMessage.payloadVariant.value}`,
+                `🚧 Received Queue Status: ${JSON.stringify(decodedMessage.payloadVariant.value)}`,
```
Unsure why this was added, it should be reverted as this would touch every piece of data received by the web app from a node.

danditomaso commented on packages/web/public/i18n/locales/en/config.json:
```diff
@@ -283,7 +283,42 @@
+      },
+      "altitude": {
+        "label": "Altitude",
+        "description": "Meters above sea level (optional)"
```
This should accept an interpolated string, as I believe if someone's localisation is set to US, it will be feet above sea level. It would look like  {{ unitOfMeasure }} in other translations you can see examples of this usage.

danditomaso reviewed (changes requested):
Overall really glad to see this has been added to the code base, it has been needed for a while. But I dont need to see some code reverted otherwise it will cause all kinds of type issues around the code base. As well the Fixed Position component must be added to the DynamicFormField and called as a `type`

dzienisz commented:
I am open to all suggestions. This is my first big PR. I can improve it next week.

dzienisz commented:
> @dzienisz Wanted to follow up and see if the fixes were going to be added to this PR?

today, sorry for delay

Commit:
feat: add interactive fixed position picker with map interface

- Created new FixedPositionPicker component with clickable map for setting device coordinates
- Added form field type for fixed position picker that appears when fixedPosition toggle is enabled
- Implemented position request functionality to retrieve current device location

Commit:
feat: display altitude unit based on user's display settings

- Added dynamic altitude unit (Meters/Feet) that respects the user's imperial/metric display preference
- Updated altitude field description to show the appropriate unit instead of hardcoded "Meters"

Commit:
refactor: replace any type with MapLayerMouseEvent in map click handler

Commit:
refactor: improve accessibility and code quality in FixedPositionPicker

- Replace hardcoded IDs with useId() hook for proper accessibility
- Use Number.isNaN() instead of isNaN() for more reliable type checking
- Add radix parameter to parseInt() and remove unnecessary fragment wrapper

Commit:
refactor: simplify fixed position picker integration

- Removed dedicated FixedPositionPicker form field type in favor of toggle's additionalContent prop
- Moved FixedPositionPicker to render conditionally within toggle field instead of as separate dynamic field
- Streamlined form field types by eliminating FixedPositionPickerFieldProps

Commit:
style: format code with consistent line breaks and import ordering

danditomaso commented:
Can you remove the large white border and the padding around this new component, and with the exception of one other comment about using standards such as {...rest} in the component it looks good to go.

danditomaso commented on packages/web/src/components/Form/FormToggle.tsx:
```diff
@@ -19,31 +25,34 @@ export function ToggleInput<T extends FieldValues>({
+          />
+        )}
+      />
+      {field.additionalContent}
```
I think this is the only spot its used? The {...rest} convention was already in the codebase and I think that convention is pretty clear in what its accomplishing.

Commit:
refactor: simplify fixed position picker container styling

Commit:
feat: disable fixed position toggle when GPS is enabled

Commit:
refactor: use ComponentRef instead of ElementRef in Switch component

danditomaso commented:
So I played with it a bit, and if I enter a lat or long greater than -90 - 90 it throws an error, or if you enter 3 digits at all, and click "set fixed position" the thrown error is caught by the top level error handler. This should really validate the input and display an error if the input isn't valid instead of throwing an error. 
<img width="790" height="235" alt="image" src="https://github.com/user-attachments/assets/b97f46e3-0336-4498-ba3b-22c82087c39f" />
<img width="1844" height="760" alt="image" src="https://github.com/user-attachments/assets/792d1f01-e938-4232-b14d-d035720b3df1" />

Commit:
refactor: replace interactive map picker with inline coordinate fields for fixed position

- Removed FixedPositionPicker component with map interface
- Added latitude, longitude, and altitude fields directly to position form
- Moved coordinate validation into PositionValidationSchema with proper min/max bounds
- Updated translation strings to include coordinate ranges and improved altitude description
- Coordinates now sent via setFixedPosition admin message on form submit when fixedPosition is enabled

Commit:
refactor: simplify toggle field by removing additionalContent prop and unused field spreading

- Removed additionalContent prop and its JSDoc documentation from ToggleFieldProps
- Removed rendering of additionalContent below toggle switch
- Cleaned up Controller render function by removing unused rest spread operator
- Renamed field destructuring to controllerField for clarity

Commit:
chore: remove unused logo SVG files

Commit:
refactor: improve fixed position handling and add position broadcast request

- Restructure onSubmit to save config before sending admin message
- Add position broadcast request after setting fixed position to immediately update display
- Add comprehensive debug logging throughout submission flow
- Extract coordinate exclusion logic earlier in submission process for clarity
- Add 1 second delay before requesting position broadcast to allow fixed position processing

dzienisz commented:
Hey @danditomaso ! 👋

I went with **Option 2** (integrate admin messages into the global save system) after evaluating both approaches.

### Why Option 2?

The main advantage is that it maintains consistency with the existing architecture. All configuration changes in the app already flow through the global save mechanism, so integrating admin messages into this same workflow makes the codebase more predictable and easier to maintain. Users also get a unified experience - one save button for everything.

### What about Option 1?

Option 1 (restoring the dedicated "Set Fixed Position" button) would have worked and been simpler to implement, but it would have created an inconsistency in the UX - users would need to understand that position coordinates require a separate save action from other settings. While this is more explicit, it breaks the pattern established by the rest of the settings interface.

### Implementation Details

The solution extends the existing change registry system to support admin messages as first-class changes:
- Admin messages are queued alongside config/module/channel changes
- They're sent **after** [commitEditSettings()](cci:1://file:///Users/home/Developer/web/packages/core/src/meshDevice.ts:594:2-609:3) to ensure proper sequencing
- The pending changes indicator includes queued admin messages
- Everything clears atomically on save

This eliminates the race condition while keeping the architecture clean and extensible for future admin message types.

Let me know if you have any questions about the implementation! 🚀

Commit:
feat: add max length constraint to latitude and longitude fields

- Set fieldLength.max to 10 for both latitude and longitude inputs
- Prevents excessive decimal precision while maintaining 7 decimal places (±1.1cm accuracy)

danditomaso commented:
I tried the version hosted on vercel (link above) and it looks like there are problems showing the fixed position fields. 
<img width="1946" height="1167" alt="image" src="https://github.com/user-attachments/assets/e215df06-19fd-4289-923c-fb5bdcb41e99" />
